harness/determined/tensorboard/base.py

```diff
@@ -76,13 +83,30 @@ def sync(
         """
         pass
 
+    def sync(
+        self,
+        selector: Callable[[pathlib.Path], bool] = lambda _: True,
+        mangler: Callable[[pathlib.Path, int], pathlib.Path] = lambda p, __: p,
+        rank: int = 0,
+    ) -> None:
+        paths = self.to_sync(selector)
+        if self.upload_thread is not None and self.upload_thread.is_alive():
+            self.upload_thread.upload(_UploadTask(paths=paths, mangler=mangler, rank=rank))
+        else:
+            self._sync_impl(paths, mangler, rank)
```

Comment:
Can we just apply the selector and the mangler up front, and pass in a list of (src, dst) pairs?

Reply:
I think we cannot do so because SharedFSTensorboardManager differs from the cloud managers (e.g. s3, Azure, GCP).

SharedFSTensorboardManager:
`mangled_path = self.shared_fs_base.joinpath(mangled_relative_path)`

Cloud:
`mangled_path = self.sync_path.joinpath(mangled_relative_path)`

Reply:
There are three steps to prepare the paths, across all subclasses

1. get relative_path
2. get mangled_relative_path
3. get mangled_path

1 and 2 are the same across all subclasses, but SharedFSTensorboardManager's step 3 is different
`mangled_path = self.shared_fs_base.joinpath(mangled_relative_path)` versus all other subclasses
`mangled_path = self.sync_path.joinpath(mangled_relative_path)`

While we can definitely process the paths until step 2, then pass to _sync_impl, I think it is a not clean that the path processing steps are partly done in sync step and partly done in async step.

I don't have a strong opinion on this though, if you really prefer applying mangler in sync step, we can do that too.

Reply:
So, I'm being spread kinda thin, and maybe I'm not understanding fully, but couldn't you just pass `(src, relative_dst)` paths in that case?

Reply:
Sure we can, I just didn't feel strongly about that approach stylistically as it breaks the path cleaning steps into two parts. I am also not against it either. I can go ahead and make the change.

Reply:
The thing I like the most is making all the calls to user-provided functions in a single-threaded context; I think having the async code be as dumb as possible improves the understandability of the system from the user's perspective.

Reply:
Yes, I agree with this principle, I am already making changes towards this end, thank you!